btree/btree.c

```diff
@@ -2357,11 +2362,55 @@ static bool ff_rec_is_valid(const struct slot *slot)
 	   _0C(val_is_valid);
 }
 
+static bool ff_iskey_smaller(const struct nd *node, int prev_key_idx,
+			int next_key_idx)
+{
+	struct ff_head          *h;
+	struct m0_btree_key      key_prev;
+	struct m0_btree_key      key_next;
+	struct m0_bufvec_cursor  cur_prev;
+	struct m0_bufvec_cursor  cur_next;
+	void                    *p_key_prev;
+	m0_bcount_t              ksize_prev;
+	void                    *p_key_next;
+	m0_bcount_t              ksize_next;
+	int                      diff;
+
+	h          = ff_data(node);
+	ksize_prev = h->ff_ksize;
+	ksize_next = h->ff_ksize;
+
+	key_prev.k_data = M0_BUFVEC_INIT_BUF(&p_key_prev, &ksize_prev);
+	key_next.k_data = M0_BUFVEC_INIT_BUF(&p_key_next, &ksize_next);
+
+	p_key_prev = ff_key(node, prev_key_idx);
+	p_key_next = ff_key(node, next_key_idx);
+
+	m0_bufvec_cursor_init(&cur_prev, &key_prev.k_data);
+	m0_bufvec_cursor_init(&cur_next, &key_next.k_data);
+	diff = m0_bufvec_cursor_cmp(&cur_prev, &cur_next);
+	if (diff >= 0)
+		return false;
+	return true;
+
+}
+
+static bool ff_expensive_invariant(const struct nd *node)
+{
+	int count = node_count(node);
+	return _0C(ergo(count > 1, m0_forall(i, count - 1,
+					     ff_iskey_smaller(node, i, i+1))));
+}
+
 static bool ff_invariant(const struct nd *node)
 {
 	const struct ff_head *h = ff_data(node);
 
-	return  _0C(h->ff_shift == segaddr_shift(&node->n_addr));
+	/* TBD: add check for h_tree_type after initializing it in node_init. */
+	return  _0C(h->ff_fmt.hd_magic == M0_FORMAT_HEADER_MAGIC) &&
+		_0C(h->ff_seg.h_node_type == BNT_FIXED_FORMAT) &&
+		_0C(h->ff_ksize != 0) && (h->ff_vsize != 0) &&
```

Comment:
either have both ksize and vsize check in same _0C or and separate _0C for vsize.